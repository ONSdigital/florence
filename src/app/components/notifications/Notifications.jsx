import React from "react";
import PropTypes from "prop-types";
import NotificationItem from "./NotificationItem";

const Notifications = ({ notifications }) => {
    if (notifications.length === 0) return null;

    return (
        <ul className="notifications" data-testid="notifications">
            {notifications.map((notification, i) => (
                <NotificationItem key={notification + i} {...notification} />
            ))}
        </ul>
    );
};

Notifications.propTypes = {
    notifications: PropTypes.array,
};

export default Notifications;
