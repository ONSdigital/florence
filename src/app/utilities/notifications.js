import { addNotification, removeNotification, toggleNotificationVisibility } from '../config/actions';
import { store } from '../config/store';

export default class notifications {

    /**
     * Updates state with new notification
     * @param {object} configuration of notification , as follows: 
     * {
     *      id [String]
     *      type [oneOf(["neutral", "warning"])]
     *      message [String]
     *      autoDismiss [Number]
     *      isDismissable [Bool]
     *      buttons [Array of objects]
     * }
     */
    static add(notification) {
        const config = {
            type: notification.type || "neutral",
            message: notification.message || "",
            id: notification.id || Date.now(),
            buttons: notification.buttons || [],
            isVisible: false
        }

        if (notification.autoDismiss && notification.autoDismiss > 0) {
            const timer = window.setTimeout(() => {
                window.clearTimeout(timer);
                this.remove(config.id);
            }, notification.autoDismiss);
        }

        if (notification.isDismissable) {
            config.buttons.push({
                text: "Close",
                onClick: function() {
                    notifications.remove(config.id)
                }
            });
        }

        store.dispatch(addNotification(config));
        
        // Set a timeout so browser doesn't try to render component without the animation
        const animationTimer = window.setTimeout(() => {
            store.dispatch(toggleNotificationVisibility(config.id));
            window.clearTimeout(animationTimer);
        }, 10);
    }

    /**
     * Removes notification from state
     * @param {string} notificationID 
     */
    static remove(notificationID) {
        store.dispatch(toggleNotificationVisibility(notificationID));

        // Set a timeout so browser doesn't try to render component without the animation
        const animationTimer = window.setTimeout(() => {
            store.dispatch(removeNotification(notificationID));
            window.clearTimeout(animationTimer);
        }, 30);
    }

}