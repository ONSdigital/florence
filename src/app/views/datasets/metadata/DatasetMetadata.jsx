import React, { Component } from 'react';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import PropTypes from 'prop-types';

import http from '../../../utilities/http';
import { errCodes } from '../../../utilities/errorCodes'
import datasets from '../../../utilities/api-clients/datasets';
import notifications from '../../../utilities/notifications';
import Modal from '../../../components/Modal';
import Select from '../../../components/Select';
import Checkbox from '../../../components/Checkbox';
import Input from '../../../components/Input';
import Card from '../../../components/Card';
import CardList from '../../../components/CardList';
import RelatedDataController from './related-content/RelatedDataController';
import {updateAllDatasets, updateActiveDataset} from '../../../config/actions';

const propTypes = {
    params: PropTypes.shape({
        datasetID: PropTypes.string
    }),
    dispatch: PropTypes.func.isRequired,
    rootPath: PropTypes.string.isRequired,
    routes: PropTypes.arrayOf(PropTypes.object).isRequired,
    datasets: PropTypes.arrayOf(PropTypes.shape({
        title: PropTypes.string.isRequired,
    })),
    dataset: PropTypes.shape({
      title: PropTypes.string.isRequired,
      description: PropTypes.string,
      keywords: PropTypes.array,
      national_statistic: PropTypes.bool,
      contacts: PropTypes.arrayOf(PropTypes.shape({
          name: PropTypes.string,
          email: PropTypes.string,
          telephone: PropTypes.string,
      })),
      qmi: PropTypes.shape({
          href: PropTypes.string,
          title: PropTypes.string,
      }),
      related_datasets: PropTypes.arrayOf(PropTypes.shape({
          href: PropTypes.string,
          title: PropTypes.string,
      })),
      publications: PropTypes.arrayOf(PropTypes.shape({
          href: PropTypes.string,
          title: PropTypes.string,
      }))
    })
}

class DatasetMetadata extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isFetchingDataset: false,
            error: null,
            showModal: false,
            modalType: "",
            relatedBulletins: [],
            relatedQMI: "",
            relatedLinks: [],
            keywords: [],
            titleInput: "",
            urlInput: "",
            editKey: "",
        };

        this.handleSelectChange = this.handleSelectChange.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
        this.handlePageSubmit = this.handlePageSubmit.bind(this);
        this.handleToggleChange = this.handleToggleChange.bind(this);
        this.handleBackButton = this.handleBackButton.bind(this);
        this.handleModalSubmit = this.handleModalSubmit.bind(this);
        this.handleCancel = this.handleCancel.bind(this);
        this.handleFormSubmit = this.handleFormSubmit.bind(this);
        this.handleAddRelatedClick = this.handleAddRelatedClick.bind(this);
        this.removeRelated = this.removeRelated.bind(this);
        this.handleEditRelatedClick = this.handleEditRelatedClick.bind(this);
        this.handleActions = this.handleActions.bind(this);
    }

    componentWillMount() {
        this.setState({isFetchingDataset: true});

        function guid() {
            function S4() {
                return (((1+Math.random())*0x10000)|0).toString(16).substring(1);
            }
            return (S4() + S4() + "-" + S4() + "-4" + S4().substr(0,3) + "-" + S4() + "-" + S4() + S4() + S4()).toLowerCase();
        }

        const APIRequests = [
            datasets.get(this.props.params.datasetID)
        ]

        if (this.props.datasets.length === 0) {
            APIRequests.push(datasets.getAll());
        }

        Promise.all(APIRequests).then(responses => {
            this.props.dispatch(updateActiveDataset(responses[0].current));
            if (this.props.datasets.length === 0) {
                this.props.dispatch(updateAllDatasets(responses[1].items));
            }

            const contact = this.props.dataset.contacts.find(details => {
                return {
                    name: details.name,
                    email: details.email,
                    telephone: details.telephone,
                }
            });

            { this.props.dataset.publications &&
              this.props.dataset.publications.map(item => {
                  const bulletin = {title: item.title, url: item.href, key: guid()}
                  const bulletins = this.state.relatedBulletins.concat(bulletin);
                  this.setState({relatedBulletins: bulletins})
              })
            }

            { this.props.dataset.related_datasets &&
              this.props.dataset.related_datasets.map(item => {
                  const link = {title: item.title, url: item.href, key: guid()}
                  const links = this.state.relatedLinks.concat(link);
                  this.setState({relatedLinks: links})
              })
            }

            if (this.props.dataset.qmi.title != "") {
                const item = this.props.dataset.qmi
                const qmi = {title: item.title, url: item.href, key: guid()}
                this.setState({relatedQMI: qmi})
            }

            this.setState({
                isFetchingDataset: false,
                description: this.props.dataset.description,
                title: this.props.dataset.title,
                keywords: this.props.dataset.keywords,
                isChecked:this.props.dataset.national_statistic,
                contactName: contact.name,
                contactEmail: contact.email,
                contactPhone: contact.telephone,
            });

          }).catch(error => {
              switch (error.status) {
                  case(403):{
                      const notification = {
                          "type": "info",
                          "message": "You do not permission to view this dataset",
                          isDismissable: true
                      }
                      notifications.add(notification);
                      break;
                  }
                  case (404): {
                      const notification = {
                          "type": "info",
                          "message": `Dataset ID '${this.props.params.datasetID}' was not recognised. You've been redirected to the datasets home screen`,
                          isDismissable: true
                      };
                      notifications.add(notification);
                      this.props.dispatch(push(`${this.props.rootPath}/datasets`));
                      break;
                  }
                  default: {
                      const notification = {
                          type: "warning",
                          message: "An unexpected error's occurred whilst trying to get this dataset",
                          isDismissable: true
                      }
                      notifications.add(notification);
                      break;
                  }
              }
              console.error("Error has occurred:\n", error);
            });
    }

    postDatasetDetails(body) {
       // Update to put to dataset api
      return http.put(`${datasets}`, body, true);
    }

    updateDatasetDetails(datasetDetailsData) {
      this.postDatasetDetails(datasetDetailsData).then(() => {
        // TO DO - Add correct path
        this.props.dispatch(push(`${this.props.rootPath}/datasets`));
      }).catch(error => {
          if (error) {
              const notification = {
                  type: 'warning',
                  isDismissable: true,
                  autoDismiss: 15000
              };
              switch (error.status) {
                  case ('UNEXPECTED_ERR'): {
                      console.error(errCodes.UNEXPECTED_ERR);
                      notification.message = errCodes.UNEXPECTED_ERR;
                      notifications.add(notification);
                      break;
                  }
                  case ('RESPONSE_ERR'): {
                      console.error(errCodes.RESPONSE_ERR);
                      notification.message = errCodes.RESPONSE_ERR;
                      notifications.add(notification);
                      break;
                  }
                  case ('FETCH_ERR'): {
                      console.error(errCodes.FETCH_ERR);
                      notification.message = errCodes.FETCH_ERR;
                      notifications.add(notification);
                      break;
                  }
              }
          }
      });

    }

    mapReleaseFreqToSelectOptions() {
        const values = [
          'Weekly', 'Monthly', 'Annually'
        ];

        return values.map(value => {
            return {
              id: value.toLowerCase(),
              name: value
            }
        });
    }

    convertKeywordsToString() {
        const keywords = this.props.dataset.keywords.map(keyword => {
            return keyword
        });
        const keywordString = keywords.join(", ");
        return keywordString;
    }

    handleSelectChange(event) {
        this.setState({
            periodicity: event.target.value
        });
    }

     handleModalSubmit(event){
       event.preventDefault();
       this.setState({showModal: false});
       this.props.dispatch(push(`${this.props.rootPath}/datasets`));
     }

    handleToggleChange(event){
      this.setState({
        isChecked: event.target.checked,
      });
    }

    handleInputChange(event) {
       const target = event.target;
       const value = target.value;
       const name = target.name;
       if (name === "add-related-content-title") {
           this.setState({titleInput: value});
           if(this.state.titleError != null) {
             this.setState({titleError: null})
           }
       } else if (name === "add-related-content-url") {
           this.setState({urlInput: value});
           if(this.state.urlError != null) {
             this.setState({urlError: null})
           }
       } else {
         this.setState({
           [name]: value
         });

       }

     }

     handleBackButton() {
         this.setState({showModal: true});
     }

     handleCancel() {
         this.setState({showModal: false});
         this.setState({modalType: ""});
         this.setState({editKey: ""});
     }

     handleAddRelatedClick(type) {
         this.setState({showModal: true});
         this.setState({modalType: type});
     }


     handleEditRelatedClick(type, key) {
         this.setState({showModal: true});
         this.setState({modalType: type});
         this.setState({editKey: key});
     }

     removeRelated(type, key) {
         function remove(arr, key) {
             arr.map((item, index) => {
                 if (item.key === key) {
                     arr.splice(index, 1);
                 }
             })
             return arr
         }

         if (type === "bulletin") {
             var bulletins = remove(this.state.relatedBulletins, key)
             this.setState({relatedBulletins: bulletins});
         } else if (type === "qmi") {
             this.setState({relatedQMI: ""});
         } else if (type === "link") {
             var links = remove(this.state.relatedLinks, key)
             this.setState({relatedLinks: links});
         }
     }

     handleActions(type, key, action){

       if(action === "edit"){
         this.setState({showModal: true});
         this.setState({modalType: type});
         this.setState({editKey: key});
       }

       if(action === "remove"){
         this.removeRelated(type, key);
       }

     }

     mapTypeContentsToCard(items){
         return (items).map(item => {
           return {
             title: item.title,
             key: item.key,
           }
         });
     }

     handleFormSubmit(event) {
         event.preventDefault();
         function guid() {
               function S4() {
                   return (((1+Math.random())*0x10000)|0).toString(16).substring(1);
               }
               return (S4() + S4() + "-" + S4() + "-4" + S4().substr(0,3) + "-" + S4() + "-" + S4() + S4() + S4()).toLowerCase();
           }

         if(this.state.titleInput == "" || this.state.urlInput == ""){
           if(this.state.titleInput == ""){
             this.setState({
                 titleError: "You must provide a title"
             });
           }
           if (this.state.urlInput == ""){
             this.setState({
                 urlError: "You must provide a url"
             });
           }
         } else {
         if (this.state.modalType === "bulletin") {
             if (this.state.editKey != "") {
                 this.removeRelated("bulletin", this.state.editKey)
             }
             const bulletins = this.state.relatedBulletins.concat({title: this.state.titleInput, url: this.state.urlInput, key: guid()});
             this.setState({relatedBulletins: bulletins});
         } else if (this.state.modalType === "qmi") {
             if (this.state.editKey != "") {
                 this.removeRelated("qmi", this.state.editKey)
             }
             const qmi = {title: this.state.titleInput, url: this.state.urlInput, key: guid()};
             this.setState({relatedQMI: qmi});
         } else if (this.state.modalType === "link") {
             if (this.state.editKey != "") {
                 this.removeRelated("link", this.state.editKey)
             }
             const links = this.state.relatedLinks.concat({title: this.state.titleInput, url: this.state.urlInput, key: guid()});
             this.setState({relatedLinks: links});
         }

         this.setState({showModal: false});
         this.setState({modalType: ""});
         this.setState({editKey: ""})
       }
     }

     handlePageSubmit(event) {

         event.preventDefault();

         const datasetDetailsData = {
           contact: {
             email: this.state.contactEmail,
             name: this.state.contactName,
             telephone: this.state.contactPhone,
           },
           description: this.state.description,
           release_frequency: this.state.periodicity,
           title: this.state.title,
           national_statistic: this.state.isChecked,
           keywords: this.state.keywords,
           qmi: {
             title: this.state.relatedQMI.title,
             href: this.state.relatedQMI.url,
           },
           publications: [...this.state.relatedBulletins],
           related_datasets: [...this.state.relatedLinks],
         }
         if (!this.state.periodicity) {
             this.setState({
                 error: "You must select the periodicity"
             });
         } else {
           this.updateDatasetDetails(datasetDetailsData);
         }
     }

    render() {
        return (
            <div className="grid grid--justify-center">
                <div className="grid__col-4">
                    <h1>Dataset details</h1>
                    <div className="margin-bottom--1">
                        &#9664; <a href="#" onClick={this.handleBackButton}>Back</a>
                    </div>
                    <p className="margin-bottom--1">This information is common across all editions of the dataset.<br/>
                        Changing it will affect all previous edition</p>
                    {this.state.isFetchingDataset ?
                        <div className="loader loader--dark"></div>
                    :
                        <div>
                            <h2 className="margin-bottom--1">Dataset</h2>
                            <div className="margin-bottom--1"><strong>ID</strong><span className="inline-block margin-left--1">{this.props.params.datasetID || "Fetching dataset ID..."}
</span></div>
                          <form className="margin-bottom--4" onSubmit={this.handlePageSubmit}>

                              <Input
                                  value={this.props.dataset.title}
                                  id="title"
                                  label="Title"
                                  onChange={this.handleInputChange}
                              />
                              <Input
                                  value={this.props.dataset.description}
                                  type="textarea"
                                  id="description"
                                  label="About this dataset"
                                  onChange={this.handleInputChange}
                              />
                              <Input
                                  value={ this.props.dataset.keywords ? this.convertKeywordsToString() : ""}
                                  id="keywords"
                                  label="Keywords"
                                  onChange={this.handleInputChange}
                              />
                              <div className="grid__col-6 margin-top--1">
                                <Checkbox
                                    isChecked={this.state.isChecked}
                                    onChange={this.handleToggleChange}
                                    label="National Statistic"
                                    id="national-statistic"
                                />
                              </div>
                              <div className="grid__col-6 margin-bottom--1">
                                  <Select
                                      contents={this.mapReleaseFreqToSelectOptions()}
                                      onChange={this.handleSelectChange}
                                      error={this.state.error}
                                      label="Release frequency"
                                  />
                              </div>
                              <h3 className="margin-bottom--1">Contact</h3>
                              <Input
                                  value={this.state.contactName}
                                  id="contactName"
                                  label="Contact name"
                                  onChange={this.handleInputChange}
                              />
                              <Input
                                  value={this.state.contactEmail}
                                  id="contactEmail"
                                  label="Contact email"
                                  onChange={this.handleInputChange}
                              />
                              <Input
                                  value={this.state.contactPhone}
                                  id="contactPhone"
                                  label="Contact telephone"
                                  onChange={this.handleInputChange}
                              />
                        <h2 className="margin-bottom--1">Related content</h2>
                        <div className="margin-bottom--1">
                            <p> These are common across all editions of the dataset. Changing them will affect all previous editions.</p>
                        </div>
                        <div className="margin-bottom--2">
                            <h3> Bulletins, articles and compendia </h3>
                            <CardList
                              contents={this.mapTypeContentsToCard(this.state.relatedBulletins)}
                              type="bulletin"
                              listActions={this.handleActions}
                              />
                            <a href="#" onClick={() => {this.handleAddRelatedClick("bulletin")}}> Add document</a>
                        </div>
                        <div className="margin-bottom--2">
                            <h3> QMI </h3>
                                { this.state.relatedQMI.title != undefined ?
                                  <ul className="list--neutral">
                                    <Card
                                      title={this.state.relatedQMI.title}
                                      keyID={this.state.relatedQMI.key}
                                      type="qmi"
                                      onEdit={this.handleActions}
                                      />
                                  </ul>
                                :
                                  <a href="#" onClick={() => {this.handleAddRelatedClick("qmi")}}> Add QMI </a>
                                }
                        </div>
                        <div className="margin-bottom--2">
                            <h3> Related links </h3>
                                <CardList
                                  contents={this.mapTypeContentsToCard(this.state.relatedLinks)}
                                  type="link"
                                  listActions={this.handleActions}
                                  />
                              <a href="#" onClick={() => {this.handleAddRelatedClick("link")}}> Add related link</a>
                        </div>
                        <button className="btn btn--positive" onClick={this.handlePageSubmit}>Save and Continue</button>
                        </form>
                    </div>
                }
                  </div>
                  {
                      this.state.showModal &&

                      <Modal sizeClass="grid__col-3">
                        {
                            this.state.modalType ?

                          <RelatedDataController
                              name="related-content-modal"
                              titleInput={this.state.titleInput}
                              urlInput={this.state.urlInput}
                              onCancel={this.handleCancel}
                              onFormInput={this.handleInputChange}
                              onFormSubmit={this.handleFormSubmit}
                              titleError={this.state.titleError}
                              urlError={this.state.urlError}
                          />
                        :
                          <div>
                          <div className="modal__header">
                              <h2>Warning!</h2>
                          </div>
                          <div className="modal__body">
                              <p>You will lose any changes by going back without saving. </p><br/>
                              <p>Click "Continue" to lose changes and go back to the previous page or
                                  click "Cancel" to stay on the current page.</p>
                          </div>
                          <div className="modal__footer">
                          <button type="button" className="btn btn--primary btn--margin-right" onClick={this.handleModalSubmit}>Continue</button>
                          <button type="button" className="btn" onClick={this.handleCancel}>Cancel</button>
                          </div>
                        </div>
                      }
                      </Modal>

                  }

          </div>
        )
    }
}

DatasetMetadata.propTypes = propTypes;

function mapStateToProps(state) {
    return {
        rootPath: state.state.rootPath,
        datasets: state.state.datasets.all,
        dataset: state.state.datasets.activeDataset
    }
}

export default connect(mapStateToProps)(DatasetMetadata);
