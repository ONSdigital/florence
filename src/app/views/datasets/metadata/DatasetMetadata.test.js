import React from 'react';
import {DatasetMetadata} from './DatasetMetadata.jsx';
import notifications from '../../../utilities/notifications';
import uuid from 'uuid/v4';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

console.error = jest.fn();

jest.mock('uuid/v4', () => () => {
    return "12345";
});

jest.mock('../../../utilities/notifications', () => {
    return {
        add: jest.fn(() => {
            //
        })
    }
});

jest.mock('../../../utilities/http', () => {
    return {
        resolve: function() {
            //
        }
    }
});

jest.mock('../../../utilities/log', () => {
    return {
        add: function() {
            //
        }
    }
});

jest.mock('../../../utilities/api-clients/datasets', () => (
    {
        get: jest.fn(() => {
            return Promise.resolve({
                next: {}
            });
        }),
        getAll: jest.fn(() => {
            return Promise.resolve({
                items: []
            })
        })
    }
));

const mockEvent = {
    preventDefault: function() {}
}

const exampleDataset = {
    current: {
        collection_id: 'fddffdfdaadf-e8ad17766a81bf589e76ef57d854945fdf0bfe546000228837aa70701506869c',
        id: '931a8a2a-0dc8-42b6-a884-7b6054ed3b68',
        links: {
            editions: {
                href: 'http://localhost:22000/datasets/931a8a2a-0dc8-42b6-a884-7b6054ed3b68/editions'
            },
            latest_version: {
                id: 'efcc4581-30b1-463b-b85b-2e2d85c4918b',
                href: 'http://localhost:22000/datasets/931a8a2a-0dc8-42b6-a884-7b6054ed3b68/editions/2016/versions/1'
            },
            self: {
                href: 'http://localhost:22000/datasets/931a8a2a-0dc8-42b6-a884-7b6054ed3b68'
            }
        },
        qmi: {
            href: 'http://localhost:8080/datasets/12345'
        },
        related_datasets: [
            {
                href: 'http://localhost:8080/datasets/6789910',
                title: 'Crime in the UK'
            },
            {
                href: 'http://localhost:8080/datasets/6789910',
                title: 'More Crime in the UK'
            }
        ],
        publications: [
            {
                href: 'http://www.localhost:8080/datasets/173849jf8j238d',
                title: 'An example publication'
            }
        ],
        methodologies: [
            {
                href: 'http://www.localhost:8080/datasets/173849jf8j238d',
                title: 'An example methodology'
            }
        ],
        next_release: 'pudding',
        keywords: [
            'keyword1',
            'keyword2'
        ],
        publisher: {},
        contacts: [{
            email: "test@email.com",
            name: "foo bar",
            telephone: "01633 123456"
        }],
        state: 'published',
        title: 'CPI',
        uri: '/economy/inflationandpricesindices/datasets/consumerpriceindices'
    }
}

const defaultProps = {
    dispatch: function(){},
    datasets: [{
        next: {
            title: "Dataset 1"
        },
        current: {
            title: "Dataset 2"
        }
    }],
    dataset: {...exampleDataset.current},
    params: {
        datasetID: "931a8a2a-0dc8-42b6-a884-7b6054ed3b68"
    }
}

test("Dataset details page matches stored snapshot", () => {
    const component = renderer.create(
      <DatasetMetadata {...defaultProps} />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test("Dataset title updates after successful fetch from dataset API on mount", async () => {
    const component = shallow(
        <DatasetMetadata {...defaultProps} />
    );

    await component.instance().componentWillMount();
    await component.update(); // update() appears to be async so we need to wait for it to finish before asserting
    expect(component.state("title")).toBe("CPI");
});

test("Correct modal type shows when user wants to add a related bulletin", async () => {
    const component = shallow(
        <DatasetMetadata {...defaultProps} />
    );
    expect(component.state("modalType")).toBe("");
    await component.instance().handleAddRelatedClick("bulletin");
    expect(component.state("modalType")).toBe("bulletin");
});

test("Removing a related QMI link updates state to be empty", async () => {
    const component = shallow(
        <DatasetMetadata {...defaultProps} />
    );
    await component.instance().componentWillMount();
    await component.update();
    expect(component.state("relatedQMI")).toBe(exampleDataset.current.qmi.href);
    component.instance().handleInputChange("");
    await component.update();
    expect(component.state("relatedQMI")).toBe("");
});

test("Handler for removing a related bulletin updates state correctly", async () => {
    const component = shallow(
        <DatasetMetadata {...defaultProps} />
    );

    await component.instance().componentWillMount();
    await component.update();
    expect(component.state("relatedBulletins").length).toEqual(1);
    expect(component.state("relatedBulletins")[0]).toMatchObject({
        key: "12345",
        title: exampleDataset.current.publications[0].title,
        url: exampleDataset.current.publications[0].href
    });
    await component.instance().handleDeleteRelatedClick("bulletin", "12345");
    await component.update();
    expect(component.state("relatedBulletins").length).toBe(0);
});

test("Related datasets are set in state correctly on mount", async () => {
    const component = shallow(
        <DatasetMetadata {...defaultProps} />
    );

    await component.instance().componentWillMount();
    await component.update();
    expect(component.state("relatedLinks").length).toEqual(2);
    component.state("relatedLinks").forEach((relatedLink, index) => {
        expect(relatedLink).toMatchObject({
            key: "12345",
            title: exampleDataset.current.related_datasets[index].title,
            url: exampleDataset.current.related_datasets[index].href
        });
    });
});

test("Changing an input value updates the state to show a change has been made", async () => {
    const component = await shallow(
        <DatasetMetadata {...defaultProps} />
    );

    await component.update();
    expect(component.state("hasChanges")).toBe(false);
    component.setState({description: "A new description"});
    expect(component.state("description")).toBe("A new description");
    expect(component.state("hasChanges")).toBe(true);
});

test("Warning modal shown when unsaved changes have been made", async () => {
    const component = await shallow(
        <DatasetMetadata {...defaultProps} />
    );

    await component.update();
    expect(component.find(".modal__header h2").exists()).toBe(false);
    expect(component.state("showModal")).toBe(false);
    component.setState({description: "A new description"});
    component.instance().handleBackButton();
    expect(component.state("showModal")).toBe(true);
    expect(component.state("modalType")).toBe("");
    expect(component.find(".modal__header h2").exists()).toBe(true);
});

test("Available release frequencies maps correctly to select element", () => {
    const component = shallow(
        <DatasetMetadata {...defaultProps} />
    );

    const validSelectContents = [
        {
            id: "weekly",
            name: "Weekly"
        },
        {
            id: "monthly",
            name: "Monthly"
        },
        {
            id: "annually",
            name: "Annually"
        }
    ]

    const createdSelectContents = component.instance().mapReleaseFreqToSelectOptions();
    expect(createdSelectContents).toEqual(expect.arrayContaining(validSelectContents));
});

test("Handle select change event updates state correctly", () => {
    const component = shallow(
        <DatasetMetadata {...defaultProps} />
    );

    expect(component.update().state("releaseFrequency")).toBe("");
    component.instance().handleSelectChange({preventDefault: ()=>{}, target: {value: "Weekly"}});
    expect(component.update().state("releaseFrequency")).toBe("Weekly");
});

test("Handle checkbox tick updates 'national statistic' state correctly", () => {
    const component = shallow(
        <DatasetMetadata {...defaultProps} />
    );

    expect(component.state("isNationalStat")).toBe(false);
    component.instance().handleToggleChange(true);
    expect(component.state("isNationalStat")).toBe(true);
    component.instance().handleToggleChange(false);
    expect(component.state("isNationalStat")).toBe(false);
});

test("Handle input change updates relevant state correctly", () => {
    const component = shallow(
        <DatasetMetadata {...defaultProps} />
    );
    const mockTitleEvent = {
        target: {
            value: "Some free text",
            name: "add-related-content-title"
        }
    }
    const mockURLEvent = {
        target: {
            value: "https://url.com",
            name: "add-related-content-url"
        }
    }

    expect(component.state("titleInput")).toBe("");
    expect(component.state("urlInput")).toBe("");

    component.instance().handleInputChange(mockTitleEvent);
    expect(component.state("titleInput")).toBe(mockTitleEvent.target.value);

    component.instance().handleInputChange(mockURLEvent);
    expect(component.state("urlInput")).toBe(mockURLEvent.target.value);
});

test("Input errors are added on submit and then removed on change of that input", () => {
    const component = shallow(
        <DatasetMetadata {...defaultProps} />
    );
    const mockTitleEvent = {
        target: {
            value: "Some free text",
            name: "add-related-content-title"
        }
    }
    const mockURLEvent = {
        target: {
            value: "https://url.com",
            name: "add-related-content-url"
        }
    }

    expect(component.state("titleInput")).toBe("");
    expect(component.state("urlInput")).toBe("");
    expect(component.update().state("titleError")).toBe("");
    expect(component.update().state("urlError")).toBe("");

    component.instance().handleFormSubmit({preventDefault: ()=>{}});
    expect(component.update().state("titleError")).not.toBe("");
    expect(component.update().state("urlError")).not.toBe("");

    component.instance().handleInputChange(mockTitleEvent);
    expect(component.state("titleInput")).toBe(mockTitleEvent.target.value);
    expect(component.update().state("titleError")).toBe(null);

    component.instance().handleInputChange(mockURLEvent);
    expect(component.state("urlInput")).toBe(mockURLEvent.target.value);
    expect(component.update().state("urlError")).toBe(null);
});

test("Handler to edit a related item updates the state with new values", async () => {
    const component = await shallow(
        <DatasetMetadata {...defaultProps} />
    );
    await component.update();
    
    expect(component.state("relatedBulletins").length).toBe(1);
    const initialRelatedBulletins = component.state("relatedBulletins");
    component.setState({titleInput: "Some new words"});
    component.instance().editRelatedLink("bulletin", "12345");
    expect(component.state("relatedBulletins").length).toBe(1);
    const editedRelatedBulletins = component.update().state("relatedBulletins");
    expect(editedRelatedBulletins[0].title).toBe("Some new words");
});

test("Handle for editing a related link opens modal with existing values", async () => {
    const component = await shallow(
        <DatasetMetadata {...defaultProps} />
    );
    await component.update();
    
    component.instance().handleEditRelatedClick("bulletin", "12345");
    expect(component.state("showModal")).toBe(true);
    expect(component.state("modalType")).toBe("bulletin");
    expect(component.state("urlInput")).toBe(exampleDataset.current.publications[0].href);
    expect(component.state("titleInput")).toBe(exampleDataset.current.publications[0].title);
});

test("Related items map to card element correctly", async () => {
    const component = await shallow(
        <DatasetMetadata {...defaultProps} />
    );
    await component.update();

    const cardProps = component.instance().mapTypeContentsToCard(component.state("relatedLinks"));
    component.state("relatedLinks").forEach((dataset, index) => {
        expect(cardProps[index]).toMatchObject({
            title: dataset.title,
            id: dataset.key
        });
    })
});

describe("Component's state maps to API request correctly when", () => {
    const component = shallow(
        <DatasetMetadata {...defaultProps} />
    );
    
    it("contact details data has been updated", () => {
        const mockState = {
            contactEmail: "foobar@email.com",
            contactPhone: "01234 567890",
            contactName: "Bob Jones"
        }
        const mockRequestContactsObject = {
            email: "foobar@email.com",
            telephone: "01234 567890",
            name: "Bob Jones"
        }
    
        component.setState(mockState);
        const componentInstance = component.instance();
        expect(componentInstance.mapStateToAPIRequest().contacts.length).toBe(1);
        expect(componentInstance.mapStateToAPIRequest().contacts[0]).toEqual(expect.objectContaining(mockRequestContactsObject));
    });
    
    it("description data has been updated", () => {
        const mockState = {
            description: "This is a stubbed description"
        }
    
        component.setState(mockState);
        expect(component.instance().mapStateToAPIRequest().description).toBe(mockState.description);
    });
    
    it("related publications data has been updated", () => {
        const mockPublicationsState = {
            relatedBulletins: [
                {
                    title: "A publication title",
                    href: "/economy/gdp/bulletins/july2016"
                },
                {
                    title: "A publication title",
                    href: "/economy/gdp/bulletins/july2016",
                    description: "Words words words and more words"
                }
            ]
        }
    
        component.setState(mockPublicationsState);
        const componentInstance = component.instance();
        expect(componentInstance.mapStateToAPIRequest().publications.length).toBe(2);
        expect(componentInstance.mapStateToAPIRequest().publications[0]).toEqual(expect.objectContaining(mockPublicationsState.relatedBulletins[0]));
        expect(componentInstance.mapStateToAPIRequest().publications[1]).toEqual(expect.objectContaining(mockPublicationsState.relatedBulletins[1]));
    });
    
    it("related links data has been updated", () => {
        const mockLinksState = {
            relatedLinks: [
                {
                    title: "GOV.UK",
                    href: "https://gov.uk"
                }
            ]
        }
    
        component.setState(mockLinksState);
        expect(component.instance().mapStateToAPIRequest().related_datasets[0]).toEqual(expect.objectContaining(mockLinksState.relatedLinks[0]));
    });
    
    it("related QMI data has been updated", () => {
        const mockQMIState = {
            relatedQMI: {
                url: "/economy/gdp/methodology/gdpqmi"
            } 
        }
        const mockRequestQMIObject = {
            href: mockQMIState.relatedQMI.url
        }
    
        component.setState(mockQMIState);
        expect(component.instance().mapStateToAPIRequest().qmi).toEqual(expect.objectContaining(mockRequestQMIObject));
    });
    
    it("national statistic data has been updated", () => {
        const mockNationalStatState = {
            isNationalStat: false
        }
    
        component.setState(mockNationalStatState);
        expect(component.instance().mapStateToAPIRequest().national_statistic).toBe(false);
    });
    
    it("keywords data has been updated", () => {
        const mockKeywordsState = {
            keywords: "foo , bar,more, two words, another"
        }
    
        component.setState(mockKeywordsState);
        expect(component.instance().mapStateToAPIRequest().keywords).toEqual(expect.arrayContaining(["foo", "bar", "more", "two words", "another"]));
    });
    
    it("title data has been updated", () => {
        const mockTitleState = {
            title: "Foobar"
        }
    
        component.setState(mockTitleState);
        expect(component.instance().mapStateToAPIRequest().title).toBe("Foobar");
    });
});