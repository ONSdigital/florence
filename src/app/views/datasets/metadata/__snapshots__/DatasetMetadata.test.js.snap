// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dataset details page matches stored snapshot 1`] = `
<div
  className="grid grid--justify-center"
>
  <div
    className="grid__col-4"
  >
    <h1>
      Dataset details
    </h1>
    <div
      className="margin-bottom--1"
    >
      â—€ 
      <a
        href="#"
        onClick={[Function]}
      >
        Back
      </a>
    </div>
    <p
      className="margin-bottom--1"
    >
      This information is common across all editions of the dataset.
      <br />
      Changing it will affect all previous edition
    </p>
    <div
      className="loader loader--dark"
    />
  </div>
  
</div>
`;
